### ISSUE TITLE
Wildlife Computers GPE3 Auto-Ingestion Pipeline Update

### ISSUE THREAD

**Issue by iwensu0313** [ioos/marine_life_data_network#117]
### Who is requesting this?

ATN DAC

### What is being requested?

Goal: Update the conditional logic in the ATN auto-ingestion pipeline to be more robust in handling different data/file types from Wildlife Computers.

Update the automatic data ingestion pipeline for Wildlife Computer trajectory data to account for GPE3 data that doesn't have the `gpe3` metadata in the XML file. Previously, that was the way we were told that we could identify if a trajectory data coming in had GPE3 data or not. However, it appears sometimes that is not the case.

### What is the requested deadline and why?

Ideally before we run into this issue again, to minimize troubleshooting and manually ingesting.

### What is the current status quo (i.e., what happens if this does not get done)?

If this is not done, we would have to manually ingest the data.

### What indicates this is done (i.e., how do we know this is complete)?

The pipeline is updated, unit tests passed, and GPE3 data will be ingested with no errors even if the XML file doesn't indicate that it has GPE3.

**Comment by MathewBiddle** [comment]
Is there any idea as to how much effort this would take? 

I see this is `in progress` is that truly the case? @conniekot do you agree that this is a priority right now?

**Comment by iwensu0313** [comment]
Yes - this has been in progress on and off since the issue was identified by Ian and Megan in ~Oct. I haven't paused this yet because it could be considered 'keeping the data ingestion pipeline maintained and running'. There is an issue with data provider not providing certain metadata in the XML that breaks the automation pipeline for some Wildlife computer trajectory data that we are ingesting. If this comes up again while we are doing the ingests, we will need to manually fix and ingest the data.

I pinged Connie in the Marine life google doc about it to check to see if we should continue or pause. The progress feels 95-98% done. The Merge Request for the fix has already been created and reviewed, but the unit tests aren't passing, so that's where we are at. Let me know!

**Comment by MathewBiddle** [comment]
mind linking to the Pull Request?

**Comment by iwensu0313** [comment]
@MathewBiddle - our ingestion pipeline is not in a public repo so I can't link to the MR. And it's not really an isolated/singular tool like `bufrtools`, `ioos_qc`, etc. While we package as much as possible into open-source tools, as prioritized, our data ingestion system is less a singular tool and more of an interconnected web of classes to handle different data, communicating w/ our docker images, data management systems, etc etc. It's less related to manipulation of the data and more of updating the logic to handle the data (and its various forms) as it comes in. If that makes sense.

**Comment by iwensu0313** [comment]
I checked w/ our team and it could be wrapped up in a ~couple days

**Comment by MathewBiddle** [comment]
I like the specificity of this issue and understand linking to the MR. Thanks, @iwensu0313.

I don't fully understand what the fix is for this issue. Could that be added here - to be added to the documentation.

**Comment by iwensu0313** [comment]
@MathewBiddle - summary of changes from our team:

Update the conditional logic in the ATN auto-ingestion pipeline to be more robust in handling different data/file types from Wildlife Computers.
1. Old/New logic
- Old logic: Triggered GPE3 ingestion based on presence of `<gpe3>` tag under `<default_location_process>` in the XML metadata file 
- New logic: Triggered GPE3 ingestion based on existence of actual `*GPE3.csv` data file
2. Enhance unit tests associated w/ the above change
3. Simplify some code logic (directly related to above) for general code hygiene

These updates will be able to handle the following scenarios:
- Pipeline receives `*GPE3.csv` data file but no `<gpe3>` tag under `<default_location_process>` in the XML metadata file. Outcome: Process as GPE3 ingestion
- Pipeline receives `*GPE3.csv` data file with `<gpe3>` tag under `<default_location_process>` in the XML metadata file. Outcome: Process as GPE3 ingestion
- Pipeline receives neither a `Locations.csv` file nor a `*.GPE3.csv` file. Outcome: No ingestion. An informative error message will indicate a lack of location data.
- Pipeline receives location data (e.g. PAT tag) only. And neither GPE3 data file nor FoieGras model exists. Outcome: Ingest satellite tag as normal. GPE3/Foie Gras ingestion will be skipped.

**Comment by MathewBiddle** [comment]
This is helpful. Thanks!

**Comment by iwensu0313** [comment]
Has been reviewed by lead engineer - waiting to push changes live until Mon. This will allow us to auto-ingest Wildlife Computer (WC) GP3 data that we've had to troubleshoot because some deployments went through and others didn't. Once this is done, some priority ingests (e.g. Skomal) will be unblocked!

**Comment by iwensu0313** [comment]
Fix has been implemented for WC auto ingestion pipeline, Skomal ingest is unblocked. Roughly 10% of ATN ingests have GPE3 data and would be relevant to this.
